### ISSUE TITLE
Simpletest semi-silently ignores wrong file paths in .tests.info file

### ISSUE THREAD

**Issue by bugfolder** [backdrop/backdrop-issues#5491]
As noted in https://github.com/backdrop/backdrop-issues/issues/5490, 

> it turned out that several tests actually never run:
> core/modules/system/tests/system.tests.info contains wrong file paths
> It's still not clear, why these tests ran in the past, but don't for several weeks. Eventually related: Issue https://github.com/backdrop/backdrop-issues/issues/5415: Gracefully handle empty or malformed .tests.info files.

In https://github.com/backdrop/backdrop-issues/issues/5415, we introduced code that checked for malformed .tests.info files and put errors into watchdog when they're detected. We can (should) add to this error-checking to verify that the file paths specified in the tests.info file are correct.

PR to follow.

**Comment by bugfolder** [comment]
The PR can be tested with the existing code base (prior to merging PR https://github.com/backdrop/backdrop/pull/3942) by:

* Enabling Simpletest
* Visiting the testing page at admin/config/development/testing
* Going to watchdog at admin/reports/dblog and examining the most recent entry.

**Comment by bugfolder** [comment]
Here's the Tugboat dblog report:

![dblog](https://user-images.githubusercontent.com/29214084/152451266-dfffa5b1-7669-469b-81ce-01ac5b9904c0.jpg)

**Comment by indigoxela** [comment]
@bugfolder many thanks for the PR, the code looks good.

I do have two concerns, though:

- The watchdog message happens only once (simpletest caches the result), so this warning may be overlooked, unless caches get flushed often.
- Besides that log message, there's now absolutely nothing that indicates a problem. Via UI tests simply don't appear, in command line runs tests aren't listed, either.

Your PR hides problems even more. For instance, in our GHA nobody would ever realize that something's broken, because what previously caused errors is now in complete stealth mode. That seems a bit dangerous to me.

**Comment by klonos** [comment]
Indeed, thank you @bugfolder ...but I'm also with @indigoxela on this: we need to be more "verbose" when there are wrong paths in test files. I'd suggest at least the following:
- when the simpletest module is enabled if test files with erroneous file paths are detected, there's an error shown in the test selection page at `admin/config/development/testing` (and perhaps also on the site status page?).
- we make it so that tests fail hard when test file paths are wrong (test files missing)

**Comment by indigoxela** [comment]
@klonos yes, that would make sense IMO :+1: 

A quick idea regarding command line: we have function [backdrop_is_cli()](https://docs.backdropcms.org/api/backdrop/core%21includes%21bootstrap.inc/function/backdrop_is_cli/1), which we can use to let tests fail for cli, but only show an error on the selection page when using UI, still suppressing fatal errors. Not sure how easy/complex that would be.

**Comment by bugfolder** [comment]
* Agreed that stronger notification is needed. Seems like any of the error states detected in `simpletest_get_all()` would warrant strong notification, not just the bad file path error.
* An easy way to highlight problems in the UI would be to post the same error message using `backdrop_set_message()` at the same time as the `watchdog()` entry. Reasonable?
* We could ensure that the check is always run when viewing the UI if we cleared the static variable in `simpletest_test_form()`, but that would slow down form loading. Is that an acceptable price to pay (considering that the vast majority of the time there won't be any errors)?
* using `backdrop_is_cli()` we could throw a failed test assertion instead of posting messages and watchdog, but there's other reasons to run backdrop from the CLI (e.g., cron calls) that wouldn't warrant a hard fail. Perhaps testing `setUp()` routine could post a global variable `$_running tests` (maybe a static variable in BackdropWebTestCase), or something like that, that could be probed to know whether tests are running.

**Comment by indigoxela** [comment]
> An easy way to highlight problems in the UI would be to post the same error message using backdrop_set_message() at the same time as the watchdog() entry. Reasonable?

Yes, makes sense. Hm... regarding the static cache - I'm not sure if I understand what you're after.

> using backdrop_is_cli() we could throw a failed test assertion instead

This might be a bit tricky (the assertion part), because the error already happens when collecting data and the test class itself isn't loaded. But we could do something easier like trowing an Exception. Then it would (should?) be visible in the GHA and also when running it locally via cli. And people could get the location and name of the failing test.info file.

```
throw new Exception('Some helpful hint with path and name here.');
```

My main concern is that currently it might appear as if "everything's green", although there's a severe problem. An exception could be enough to indicate "a problem needs to get addressed". There may be other opinions, though.

> ...but there's other reasons to run backdrop from the CLI (e.g., cron calls) that wouldn't warrant a hard fail

I don't actually think that a cron run triggers function simpletest_test_get_all().

**Comment by bugfolder** [comment]
> I don't actually think that a cron run triggers function simpletest_test_get_all().

Doi!

**Comment by bugfolder** [comment]
> regarding the static cache - I'm not sure if I understand what you're after.

Sorry, not the static cache, the db cache (the `cache()->set('simpletest', $groups)` line in `simpletetest_get_all()`). If you refresh the testing listing page without clearing the cache, the test code doesn't run because the page is using the cached values.

**Comment by indigoxela** [comment]
> Doi!

Huh? :upside_down_face:

> not the static cache, the db cache (the cache()->set('simpletest', $groups)

Ah, OK, now I get it. :smile: Hm... I'd prefer if we find a solution without removing that cache. Not sure what would be smart here.

**Comment by bugfolder** [comment]
>Doi!
>Huh? 

Sorry, an Americanism. It means, roughly, "I should have realized that if I had two neurons to rub together." 😉

**Comment by bugfolder** [comment]
>  I'd prefer if we find a solution without removing that cache. Not sure what would be smart here.

Maybe we start by trying it without explicitly clearing the cache. If people are doing development, there's probably some cache-clearing going on, and that's when they'd most need to catch typos in the file path (or other malformations).

**Comment by bugfolder** [comment]
Well, that was suprising. If I just throw an exception in `simpletest_get_all()`, the attempted test run fails, but shows up green in the PR. Not what we want.

![test](https://user-images.githubusercontent.com/29214084/152616492-b17d8b7a-d6b0-4e0a-81d5-335d3a71b7ad.jpg)

**Comment by bugfolder** [comment]
So, how to turn that unhandled exception into a GHA failure exit code?

**Comment by indigoxela** [comment]
Huh? Seems like although there's an exception, the exit code of run-tests.sh is still 0? :confused: And GHA interprets this 0 as success...

Can we try something like that instead?

```
print "$error\n";
exit(1);
```
Just to see how that shows up in the action.

**Comment by bugfolder** [comment]
Much better. (I also eliminated the HTML formatting (`@uri` versus `%uri`) from the error messages to make them usable both in the UI and as GHA error messages.)

![test results](https://user-images.githubusercontent.com/29214084/152647795-ef7a43ed-0034-4938-881e-fc97e74aeba1.jpg)

**Comment by bugfolder** [comment]
I've now removed the intentional error (in `user.tests.info`) from the PR.

**Comment by indigoxela** [comment]
@bugfolder many thanks for updating. And removing the formatting makes messages much easier to read, when running on command line (both, local/manual and in GHA).

No more hidden-away errors - wrong file paths in info files are now loud enough. Works for me. :+1:

I had another idea re errors - doing it like function assert and put the individual test failures in the database, but didn't dig deeper. IMO that would be a waste of resources, as at that point we already know, the tests have to fail. Running the remaining (still correct) tests would be a waste of time and CPU.

**Comment by bugfolder** [comment]
@indigoxela, @klonos, either of you up for a code review to get this to RTBC?

**Comment by BWPanda** [comment]
I'm not sure about the PR... In some places `$error` is a plain string, in others it uses the `t()` function, and in yet others it uses `format_string()`. Finally, at the end, it seems that regardless of how it was set above, it's then passed through `format_string()`...

Am I reading it wrong?

**Comment by indigoxela** [comment]
> ... plain string, ... the t() function, ... format_string()

Not to forget format_plural(). :wink:

TBH I never completely understood, when to translate a string in simpletest and when not.

Re plain string vs. format_string - the latter is used when it's necessary to insert variables in a string. It won't get translated.
format_plural does translate, though.

**Comment by argiepiano** [comment]
I find this posting very helpful. In a nutshell, only use t when _testing_ translations, or when you want to match UI text that may be translated.

https://www.drupal.org/docs/7/testing/simpletest-testing-tutorial-drupal-7#t

**Comment by bugfolder** [comment]
So, although Simpletest module is all about testing, the error messages in this PR are not "messages that result from running a test," but rather are messages that the user sees in the UI (via `backdrop_set_message()` &/or `watchdog()`) when they're deciding what tests to run. So the messages in this PR should be translated.

And it's a little confusing because `format_string()` explicitly doesn't run its string arguments through `t()`, but `format_plural()` does. So in this PR, a string destined for `format_string()` should have been run through `t()`, but we can (and should) pass "raw" strings to `format_plural()`, which will do its own `t()`-ifying.

With all that being said, @BWPanda did indeed catch an inconsistency: I missed the very first case (it needed `t()`), now fixed in the PR.

**Comment by bugfolder** [comment]
@indigoxela, @klonos, @argiepiano, any of you up for a code review of this?

**Comment by indigoxela** [comment]
@bugfolder I've left some questions. It's not about newly introduced things, but pre-existing.

**Comment by bugfolder** [comment]
@indigoxela, thanks for the review! I added some explanation to your first comment; more details are in [my comment above](https://github.com/backdrop/backdrop-issues/issues/5491#issuecomment-1172944489). 

I closed and reopened the PR, and now [PHPCS is complaining about entirely unrelated files](https://github.com/backdrop/backdrop/actions/runs/3363918624/jobs/5577673674) (entityreference module), so I'm going to ignore that for now.

**Comment by indigoxela** [comment]
@bugfolder I opened a [separate issue](https://github.com/backdrop/backdrop-issues/issues/5848) for the phpcs behavior.

**Comment by bugfolder** [comment]
I rebased the PR. I also added some comments and made some changes to make the delayed substitution of variables more explicit and, therefore, hopefully clearer. And put back the '%'-style placeholders that were originally there, which also improves clarity a bit, I think.

**Comment by bugfolder** [comment]
@indigoxela, I think the PHPCS issues are being dealt with in your new issue (thanks!); would you (&/or @argiepiano)  take a look at the new PR on this issue and see if the changes address your concerns about splitting up the variable substitutions (they're now applied all in one place)?

**Comment by jenlampton** [comment]
I did a quick review of the PR and found 2 issues, one is just a comment that should be removed, the other is a use of `format_string()` where I think we needed `t()` instead.

**Comment by bugfolder** [comment]
I've replied to the two comments. Per the request, I removed the code comments of explanation for where we're using `t()` and why. And then, per the second comment, we should not use `t()` on already translated strings. That is,

```
$foo = t('a string');
$bar = t($foo);
```

should not be done.

**Comment by klonos** [comment]
> @indigoxela, I think the PHPCS issues are being dealt with in your new issue (thanks!);

If we want to quickly get rid of the PHPCS errors, we can review and merge #5888.

**Comment by bugfolder** [comment]
Now rebased as https://github.com/backdrop/backdrop/pull/4281.

**Comment by klonos** [comment]
@bugfolder thanks for updating the PR 👍🏼 

There's some odd usage of `t()` and the `$variables` array. I have seen this practice before, where the array of substitutions to be passed to `t()` is compiled beforehand, but then all instances of `t()` always include that variable (which is not what you are doing in your PR). In your PR you call `t()` but never pass the array of substitutions to it as a parameter. I see that you are doing that later, when you finally call `backdrop_set_message()` and `watchdog()`, but not sure if that works or makes sense. I believe that while you are not passing any substitutions to `t()`, then `$error` should remain a plain string.

I see what you are trying to do, but 🤔 🤔 🤔

**Comment by klonos** [comment]
... @bugfolder I've added a series of comments and suggestions re usage of `t()` ...sorry if they are not exactly "sequential", but if you look at the them in the files tab of the PR from top top bottom, they should be making sense.

**Comment by bugfolder** [comment]
Hi @klonos, you wrote:

> I see that you are doing that later, when you finally call backdrop_set_message() and watchdog(), but not sure if that works or makes sense.

Well, it does work, and I would claim that it makes sense, too. The baroque structure is due to the following semi-conflicting issues governing the final posting of the error messages:

* `backdrop_set_message` wants to see a translated message with variables already substituted;
* `watchdog` wants to see an untranslated message with separate variables;
* The message and variables going to both of those calls varies with the circumstances.

As I mentioned in the comment above, you should not pass a variable to `t()`. So this call in your suggested changes is wrong:

```
backdrop_set_message(t($error, $variables), 'error');
```

See https://docs.backdropcms.org/api/backdrop/core%21includes%21bootstrap.inc/function/t/1.

You do go on to say 

>...but if you already t() $error as in my suggestions above in this PR, then you can simply do this here:

```
backdrop_set_message($error, 'error');
```

But if you do that (`$error` is already translated AND variables substituted in), then `$error` is no longer correctly formatted for the call to `watchdog`—it no longer contains placeholders for `$variables`.

While `t()` performs the dual functions of (a) translation, (b) substituting in variables, there's no requirement for it to do both within `t()` as long as they both eventually get done. It's perfectly functional (and so, I would think, legal) to do this:

```
$msg = t('My name is @name');
$msg = format_string($msg, array('@name' => 'Robert'));
```

That's what I'm doing here. Note that the only thing `t($string, $args)` uses its `$args` argument for is the call to `format_string()` at the very end of `t()`, so following up a call to `t()` with no args with a call to `format_string()` with args is doing exactly what `t()` would have done in the first place.

**Comment by argiepiano** [comment]
@bugfolder thanks for the detailed explanation. 

Your approach does make sense to me. I see the benefit of delaying the substitution of placeholders until the end, and yes, it's possible to pass a string with a placeholder to `t()` without providing an array of substitutions. The string gets translated, but the placeholder will remain as-is without substitution.  The placeholder remains intact, and gets substituted later with `format_string()`.

I'm inclined to mark this back to `needs code review`. @klonos, any objection?

**Comment by quicksketch** [comment]
I had to read this over a few times. Reading the PR alone, it seemed @klonos was right, but reading the explanation above, I see what @bugfolder is doing. It's non-conventional but I don't see any problems it would cause.

I think some inline documentation would help a lot here, both when a string is initially passed to `t()` without variables (which looks wrong) and again when `format_string()` is called at the bottom to put in the variable substitutions.

**Comment by bugfolder** [comment]
> I think some inline documentation would help a lot here...

Hmm. I added the inline documentation in https://github.com/backdrop/backdrop/pull/4281/commits/688d88d13cd1e22a7ed2751e873eadd2d2312bca. It said:

```
        // Error messages will get translated as we create them below, but we
         // won't substitute in the variables in the calls to t() because
         // watchdog() wants to see the variables as a separate argument.
         // Instead, accumulate them in $variables as needed.
```

Then @jenlampton requested that they be removed in [the comment above](https://github.com/backdrop/backdrop-issues/issues/5491#issuecomment-1335916006), so I took it out in https://github.com/backdrop/backdrop/pull/4281/commits/926d489db4489eadae3aa2259f4cb1378e6e9450.

I could put it back, but perhaps we could get agreement between lamptons on whether to include it? Or maybe suggest a different acceptable wording?

**Comment by quicksketch** [comment]
I don't think @jenlampton understood what was happening either, based on https://github.com/backdrop/backdrop/pull/4249#pullrequestreview-1203182937; where in the same code review, she asked why `format_string()` was used instead of `t()`. I think your documentation was pretty good, but maybe add that the same message will also be displayed in `backdrop_set_message()`.

How about this:

> Error messages are translated as they are created below, but the variable substitutions are not passed in. This so the same error message can to be output to watchdog, which needs the variables separately, and to backdrop_set_message(), which needs the string with substitutions already made.

**Comment by bugfolder** [comment]
Both comments added (slightly tweaked, but basically what you said). Thanks!

**Comment by klonos** [comment]
With inline comments added as requested, this is back to RTBC once again. Thanks @bugfolder 🙏🏼

**Comment by quicksketch** [comment]
Thanks for your persistence on this @bugfolder! I merged https://github.com/backdrop/backdrop/pull/4281 into 1.x and 1.23.x. 

https://github.com/backdrop/backdrop/commit/5bfc51f8be8c084503f37a9ba4c90d687d99daa3 by @bugfolder, @indigoxela, @argiepiano, @klonos, @quicksketch, and @jenlampton.
